HIVE-24484: Upgrade Hadoop to 3.3.1 And Tez to 0.10.2

ayushtkn opened the pull request:
Upgrade to Hadoop-3.3.1 & Tez to 0.10.2

steveloughran commented:
i have a 3.3.3 RC1 coming out this week, if that helps

ayushtkn commented:
My Last run here had 4 errors, of which I think I have fixed 3 more. The one remaining is some XML parsing error, which I think might get auto resolved or may be an after affect.

@kgyrtkirk I have sorted the JLine issue here as well, which you told in the previous PR. Can you give a check once

kgyrtkirk commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java:
```diff
@@ -123,57 +122,24 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
-                    + "'hive.exec.copyfile.maxsize'='0', 'distcp.options.skipcrccheck'='')")
-            .run("use " + replicatedDbName)
-            .run("repl status " + replicatedDbName)
-            .verifyResult(tuple.lastReplicationId)
```
wasn't this the expected behaviour?

ayushtkn replied:
DistCp itself fails, It is running with hive.repl.add.raw.reserved.namespace and you can't copy if the key is not present on target cluster. Earlier I converted this to a failure case test, but then the next iteration fails which is without hive.repl.add.raw.reserved.namespace because the last load wasn't successful, so I kept the success case

kgyrtkirk commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationOnHDFSEncryptedZones.java:
```diff
@@ -123,57 +122,24 @@ public void targetAndSourceHaveDifferentEncryptionZoneKeys() throws Throwable {
               put(HiveConf.ConfVars.REPLDIR.varname, primary.repldDir);
             }}, "test_key123");
 
-    List<String> dumpWithClause = Arrays.asList(
```
seems like a testcase was removed; I wonder if this it not supported anymore ? ...and why are we removing this case in the scope of a hadoop upgrade?

ayushtkn replied:
Same as above:
DistCp itself fails, It is running with hive.repl.add.raw.reserved.namespace and you can't copy if the key is not present on target cluster. Earlier I converted this to a failure case test, but then the next iteration fails which is without hive.repl.add.raw.reserved.namespace because the last load wasn't successful, so I kept the success case

I am not sure how it was working before, but with todat raw type the key should be there, earlier this test was also single instance, while fixing it became 2 instances....

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/io/RecordReaderWrapper.java:
```diff
@@ -69,7 +70,14 @@ static RecordReader create(InputFormat inputFormat, HiveInputFormat.HiveInputSpl
+    try {
+     innerReader = inputFormat.getRecordReader(split.getInputSplit(), jobConf, reporter);
+    } catch (InterruptedIOException iioe) {
+      // If reading from the underlying record reader is interrupted, return a no-op record reader
```
Answer is here & this does fixes a couple of test so I picked it:
https://github.com/apache/hive/pull/1742/files#r674896581

ayushtkn replied:
Does this log line work:
```
      LOG.info("Interrupted while getting the input reader for {}", split.getInputSplit());
```
For 2nd I suppose it can get interrupted on any abort, not very sure, do you have any suggestions

ayushtkn commented on itests/pom.xml:
```diff
@@ -352,6 +352,12 @@
         <artifactId>hadoop-yarn-client</artifactId>
         <version>${hadoop.version}</version>
+        <exclusions>
+          <exclusion>
+            <groupId>org.jline</groupId>
+            <artifactId>jline</artifactId>
+          </exclusion>
```
Just tried. Started a Hive cluster with derby, init hive db, started HS2, then beeline.
show databases;
show tables;
create table emp(id int)
 insert into emp values (1),(2),(3),(4);
select * from emp;
show create table emp;

Jline was used in Beeline, I think it should have broken that. Let me know what else can be tested.

ayushtkn commented on ql/src/java/org/apache/hadoop/hive/ql/security/authorization/StorageBasedAuthorizationProvider.java:
```diff
@@ -178,8 +178,7 @@ public void authorize(Database db, Privilege[] readRequiredPriv, Privilege[] wri
     try {
-      if (MetaStoreServerUtils.checkUserHasHostProxyPrivileges(user, conf,
-              HMSHandler.getIPAddress())) {
+      if (MetaStoreServerUtils.checkUserHasHostProxyPrivileges(user, conf, HMSHandler.getIPAddress())) {
```
Max LineLength allowed I guess is 120?
https://github.com/apache/hive/blob/master/checkstyle/checkstyle.xml#L159-L160

ayushtkn commented on standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:
```diff
@@ -9361,7 +9362,8 @@ public NotificationEventsCountResponse get_notification_events_count(Notificatio
     if (!HiveMetaStore.isMetaStoreRemote() ||
-        !MetastoreConf.getBoolVar(conf, ConfVars.EVENT_DB_NOTIFICATION_API_AUTH)) {
+        !MetastoreConf.getBoolVar(conf, ConfVars.EVENT_DB_NOTIFICATION_API_AUTH) || conf.getBoolean(HIVE_IN_TEST.getVarname(),
+        false)) {
```
It is covered via test in TestReplicationScenarios#testAuthForNotificationAPIs
This method is also used mostly in replication context only I suppose for getting NotificationLog entries...

ayushtkn commented on standalone-metastore/pom.xml:
```diff
@@ -227,6 +227,10 @@
         <artifactId>hadoop-mapreduce-client-core</artifactId>
         <version>${hadoop.version}</version>
         <exclusions>
+          <exclusion>
+            <groupId>org.jline</groupId>
+            <artifactId>jline</artifactId>
+          </exclusion>
```
yeps, the best answer is to upgrade Jline, which was stuck. So, I thought to upgrade Hadoop that shouldn't block if possible, we are already on 3.1.0 which died long back

ayushtkn commented on storage-api/src/java/org/apache/hadoop/hive/common/ValidReadTxnList.java:
```diff
@@ -18,10 +18,10 @@
 package org.apache.hadoop.hive.common;
 
-import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.StringUtils;
```
Code doesn't compile with this. It is already marked as banned import, guess the logic has flaw.
https://github.com/apache/hive/blob/master/pom.xml#L1529

The dependency was getting pulled in from Hadoop & now it isn't there, so I have to change it to make it compile

ayushtkn commented on streaming/src/test/org/apache/hive/streaming/TestStreaming.java:
```diff
@@ -1317,6 +1318,11 @@ public void testTransactionBatchEmptyCommit() throws Exception {
 
+  /**
+   * Starting with HDFS 3.3.1, the underlying system NOW SUPPORTS hflush so this
+   * test fails.
```
Sure, I have removed the exception assertion. Kept the reason as is.
Just for code context, why HFlush support gets rid of the exception
```
            if (!out.hasCapability(StreamCapabilities.HFLUSH)) {
              throw new ConnectionError(
                  "The backing filesystem only supports transaction batch sizes of 1, but " + transactionBatchSize
                      + " was requested.");
            }
```

ayushtkn commented on common/pom.xml:
```diff
@@ -195,6 +194,11 @@
+    <dependency>
+      <groupId>org.fusesource.jansi</groupId>
+      <artifactId>jansi</artifactId>
+      <version>2.3.4</version>
```
Done

ayushtkn commented:
The only test failure here is intermittent. Have answered/addressed all the comments. 
One test I have disabled, firstly it was not failing itself but was corrupting the XML, it wasn't functional, but some test infra stuff and relying on hadoop ls command, for which output was intermittently changing. Not a good test to have either.
For record it is this:
http://ci.hive.apache.org/job/hive-precommit/job/PR-3279/7/testReport/junit/TEST-org.apache.hadoop.hive.cli.split0.TestMiniLlapLocalCliDriver/xml/_failed_to_read_/

Can't decode the failure reason here, it was that broken test which was causing this. 
If everything is good here, and only this test block. I will have a followup jira and figure this test out with the original author of the test.

I have tried basic stuff with Hive-On-MR.

ayushtkn commented on hcatalog/core/src/test/java/org/apache/hive/hcatalog/mapreduce/TestHCatMultiOutputFormat.java:
```diff
@@ -315,18 +320,19 @@ public void testOutputFormat() throws Throwable {
-        fs.getFileStatus(partitionFile.getParent().getParent()).getPermission(),
-        new FsPermission(tablePerms[i]));
+      final Path partitionFile = new Path(warehousedir + "/" + tableNames[i] + "/ds=1/cluster=ag/part-m-00000");
+      final Path grandParentOfPartitionFile = partitionFile.getParent();
```
Changed. I picked it as is from the previous PR, when I saw this test failing :-)

steveloughran commented:
> Can't decode the failure reason here, it was that broken test which was causing this

test not fully flushing/closing xml file before trying to read it?

changes look ok to me; the move in 3.3.3 to reload4j might add some exclusion complications if hive is declaring its own logging classes.

ayushtkn commented:
>test not fully flushing/closing xml file before trying to read it?

Looks like some maven issue only, it tries to compute diff between the generated query output file & the stored query output file, if the diff is empty, the test is said to be passing else failed.
else it puts the diff in the xml, Guess diff has some character which is messing up the xml structure, I need to further investigate though....

> the move in 3.3.3 to reload4j might add some exclusion complications if hive is declaring its own logging classes.

I just pushed a commit upgrading to 3.3.3, I haven't tested the distro, but the compilation & ran one test. Do I need to exclude reload4j from every hadoop dependency?
If it creates some runtime issues, I think I am happy moving from 3.1.0 to 3.3.2 and rest wait for 3.4.0. Even if I exclude next time when I move to 3.4.x or above I need to take those changes back, right?

steveloughran commented:
excluding reload4j is harmless on versions without those artifacts, so safe to add and not worry too much. 

bad xmll can happen if the test reporter doesn't escape test names properly and you've managed to get some invalid xml in there. do you have any parameterized tests? check how the strings are created, as they get included. CI tools generally aren't paranoid enough about test method names as historically it was only a java method name

steveloughran commented:
LGTM. spark has gone up to the same version last week, incidentally

abstractdog commented on ql/src/test/results/clientpositive/llap/acid_table_directories_test.q.out:
```diff
@@ -163,13 +170,6 @@ POSTHOOK: Input: default@acidparttbl@p=200
 ### ACID DELTA DIR ###
 ### ACID DELTA DIR ###
 ### ACID DELTA DIR ###
-#### A masked pattern was here ####
```
in case of a hive patch, usually, I don't care about result ordering change, however, this is a hadoop upgrade, this is not expected, can we explain this change?

ayushtkn replied:
[HADOOP-12502](https://issues.apache.org/jira/browse/HADOOP-12502) is the culprit and earlier listing was sorted now it isn't so the Ls -R output will change intermittently so we can't have this test only hence disabled

kgyrtkirk commented:
I've deployed
* hadoop-3.3.3
* tez 0.10.1
* hive from the PR
running a simple insert failed with:
```
Caused by: java.lang.NoSuchMethodError: org.eclipse.jetty.server.session.SessionHandler.getSessionManager()Lorg/eclipse/jetty/server/SessionManager;
	at org.apache.hadoop.http.HttpServer2.initializeWebServer(HttpServer2.java:569)
	at org.apache.hadoop.http.HttpServer2.<init>(HttpServer2.java:550)
	at org.apache.hadoop.http.HttpServer2.<init>(HttpServer2.java:117)
	at org.apache.hadoop.http.HttpServer2$Builder.build(HttpServer2.java:425)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.build(WebApps.java:341)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:432)
	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:428)
	at org.apache.tez.dag.app.web.WebUIService.serviceStart(WebUIService.java:94)
	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
	at org.apache.tez.dag.app.DAGAppMaster$ServiceWithDependency.start(DAGAppMaster.java:1800)
	at org.apache.tez.dag.app.DAGAppMaster$ServiceThread.run(DAGAppMaster.java:1821)
2022-05-31 09:17:19,422 [INFO] [shutdown-hook-0] |app.DAGAppMaster|: DAGAppMasterShutdownHook invoked
```
maybe I've missed something - but it seems like the tez dagappmaster has issues running with the jetty because of hadoop 3.3.3

I've these settings:
```
tez/tez-site/tez.lib.uris ${fs.defaultFS}/apps/tez/tez.tar.gz
tez/tez-site/tez.use.cluster.hadoop-libs true
```

@abstractdog is hadoop-3.3.3 supported with tez-0.10.1?

abstractdog commented:
> I've deployed
> 
> * hadoop-3.3.3
> * tez 0.10.1
> * hive from the PR
>   running a simple insert failed with:
> 
> ```
> Caused by: java.lang.NoSuchMethodError: org.eclipse.jetty.server.session.SessionHandler.getSessionManager()Lorg/eclipse/jetty/server/SessionManager;
> 	at org.apache.hadoop.http.HttpServer2.initializeWebServer(HttpServer2.java:569)
> 	at org.apache.hadoop.http.HttpServer2.<init>(HttpServer2.java:550)
> 	at org.apache.hadoop.http.HttpServer2.<init>(HttpServer2.java:117)
> 	at org.apache.hadoop.http.HttpServer2$Builder.build(HttpServer2.java:425)
> 	at org.apache.hadoop.yarn.webapp.WebApps$Builder.build(WebApps.java:341)
> 	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:432)
> 	at org.apache.hadoop.yarn.webapp.WebApps$Builder.start(WebApps.java:428)
> 	at org.apache.tez.dag.app.web.WebUIService.serviceStart(WebUIService.java:94)
> 	at org.apache.hadoop.service.AbstractService.start(AbstractService.java:194)
> 	at org.apache.tez.dag.app.DAGAppMaster$ServiceWithDependency.start(DAGAppMaster.java:1800)
> 	at org.apache.tez.dag.app.DAGAppMaster$ServiceThread.run(DAGAppMaster.java:1821)
> 2022-05-31 09:17:19,422 [INFO] [shutdown-hook-0] |app.DAGAppMaster|: DAGAppMasterShutdownHook invoked
> ```
> 
> maybe I've missed something - but it seems like the tez dagappmaster has issues running with the jetty because of hadoop 3.3.3
> 
> I've these settings:
> 
> ```
> tez/tez-site/tez.lib.uris ${fs.defaultFS}/apps/tez/tez.tar.gz
> tez/tez-site/tez.use.cluster.hadoop-libs true
> ```
> 
> @abstractdog is hadoop-3.3.3 supported with tez-0.10.1?

thanks @kgyrtkirk for trying this out, I've just created TEZ-4420, as tez is on hadoop 3.3.1, and I'm not sure about compatibility

checked tez.tar.gz contents before and after bump and I got:

```
hadoop 3.3.1

tar tf tez-dist/target/tez-0.10.2-SNAPSHOT.tar.gz | grep jetty

lib/jetty-server-9.4.40.v20210413.jar
lib/jetty-http-9.4.40.v20210413.jar
lib/jetty-io-9.4.40.v20210413.jar
lib/jetty-util-9.4.40.v20210413.jar
lib/jetty-servlet-9.4.40.v20210413.jar
lib/jetty-security-9.4.40.v20210413.jar
lib/jetty-util-ajax-9.4.40.v20210413.jar
lib/jetty-webapp-9.4.40.v20210413.jar
lib/jetty-xml-9.4.40.v20210413.jar
lib/jetty-client-9.4.40.v20210413.jar

hadoop 3.3.3

tar tf tez-dist/target/tez-0.10.2-SNAPSHOT.tar.gz | grep jetty

lib/jetty-server-9.4.43.v20210629.jar
lib/jetty-http-9.4.43.v20210629.jar
lib/jetty-io-9.4.43.v20210629.jar
lib/jetty-util-9.4.43.v20210629.jar
lib/jetty-servlet-9.4.43.v20210629.jar
lib/jetty-security-9.4.43.v20210629.jar
lib/jetty-util-ajax-9.4.43.v20210629.jar
lib/jetty-webapp-9.4.43.v20210629.jar
lib/jetty-xml-9.4.43.v20210629.jar
lib/jetty-client-9.4.43.v20210629.jar
``` 

tez packs jetty from hadoop, so hadoop upgrade means jetty upgrade too, so there is a chance that an old tez.tar.gz can clash with new hadoop

I've just uplodaded the new tez.tar.gz, is there a chance you can give it a try?
https://drive.google.com/file/d/18RMfh40s6kKdFt77E7HpS-j4EJhd2DKi/view?usp=sharing

steveloughran commented:
jetty upgrade came in https://issues.apache.org/jira/browse/HADOOP-17796 & https://github.com/apache/hadoop/pull/3208  some security advisories there so it is probably better to deal with the change than try and stick to the older version. sorry

steveloughran commented:
> I would guess the directory listing order might have changed...

shouldn't have AFAIK

TheCodeTracer commented:
I am sorry if its the incorrect forum but since it's tracking Hadoop 3.3 upgrade on Hive, I just wanted to confirm if there is a possibility that webhcat might not work with Hadoop 3.3.x yet (https://issues.apache.org/jira/browse/HIVE-24083 ). Thanks a lot. Right now, it does seem that WebHCat tests are breaking anyway due to a different issue (https://issues.apache.org/jira/browse/HIVE-26286)

ayushtkn commented:
These bunch of failures looks like due to Hadoop only, Since they are passing with my Hadoop Upgrade PR. We can merge that first, then rebase this and merge post that?

abstractdog commented:
> These bunch of failures looks like due to Hadoop only, Since they are passing with my Hadoop Upgrade PR. We can merge that first, then rebase this and merge post that?

if precommit tests cannot pass cleanly without both upgrades (hadoop+tez), we should commit those together also
(because we cannot even revert them later one by one if needed), in this case, jira title and commit message might want to contain hadoop and tez upgrade too

ayushtkn commented:
Thanx @abstractdog We have a green build here, so I have updated the PR title, while merging they will get squashed automatically :-)

abstractdog commented:
> Thanx @abstractdog We have a green build here, so I have updated the PR title, while merging they will get squashed automatically :-)

makes sense, thanks! I wish we could merge this now, hopefully we can release tez in 2 weeks

I can see the PTF boolean patch in the commits, it's not intentional I guess

suryanshagnihotri commented:
@ayushtkn Did you not face this compilation error. I do not see any change in `FairSchedulerShim.java`. I compiled hive 3.1.2 with hadoop 3.3.1.
`Compilation failure
[ERROR] /Users/suryansh/Documents/BDS/apache_hive/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java:[31,68] org.apache.hadoop.yarn.server.resourcemanager.scheduler.fair.QueuePlacementPolicy is not public in org.apache.hadoop.yarn.server.resourcemanager.scheduler.fair; cannot be accessed from outside package`

ayushtkn commented:
@suryanshagnihotri If this was the case, the build would have failed, which it didn't. So, we are cool here. Nothing pending here apart from awaiting an official Tez release here..

suryanshagnihotri commented:
@ayushtkn your base branch is does not seems to be based on 3.1.x release branch. It does not have `FairSchedulerShim.java` https://github.com/ayushtkn/hive/tree/HIVE-24484/shims/scheduler.
It fails if base branch is cut from 3.1.x...
Is there any plan to add support in 3.1.x version of hive?

ayushtkn commented:
Nopes, not chasing that branch

steveloughran commented:
it's targeting 3.3.3+; there is actually a 3.3.4 RC coming out today with specific changes to assist tez (HADOOP-18332).

sujith71955 commented:
> @ayushtkn your base branch is does not seems to be based on 3.1.x release branch. It does not have `FairSchedulerShim.java` https://github.com/ayushtkn/hive/tree/HIVE-24484/shims/scheduler. It fails if base branch is cut from 3.1.x...
> 
> Check https://github.com/apache/hive/blob/release-3.1.3-rc3/shims/scheduler/src/main/java/org/apache/hadoop/hive/schshim/FairSchedulerShim.java#L31 makes reference to `QueuePlacementPolicy` which is package protected final class https://github.com/apache/hadoop/blame/release-3.3.3-RC1/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/QueuePlacementPolicy.java#L54
> 
> Is there any plan to add support in 3.1.x version of hive?

But Hive 3.1.x version is not very old and 4.x still looks like in alpha so we may not able to upgrade.  so with this PR still we have compatibility issues with Hive 3.x version.  any suggestions? thanks
cc @sankarh

ayushtkn commented:
Got a new Test failure due to jetty. TestSSL, Fixed in the latest commit.

abstractdog commented on ql/src/test/queries/clientpositive/acid_table_directories_test.q:
```diff
@@ -1,3 +1,5 @@
+--! qt:disabled:disabled Tests the output of LS and that changes, Not a functional test, just adds some masking logic
```
please include hadoop upgrade as the cause in this comment, otherwise looks good to me

abstractdog reviewed (approved):
minor note, otherwise LGTM +1

abstractdog reviewed (approved):
LGTM +1

ayushtkn commented:
Merged. Thanx @abstractdog, @kgyrtkirk and @steveloughran for helping with reviews. :-)

abstractdog commented:
also, let me grab the opportunity to thank @belugabehr who put enormous efforts into the hadoop upgrade in the early days!
